compiler/rustc_middle/src/query/mod.rs

```diff
@@ -833,13 +833,14 @@ rustc_queries! {
     /// additional requirements that the closure's creator must verify.
     query mir_borrowck(key: LocalDefId) -> &'tcx mir::BorrowCheckResult<'tcx> {
         desc { |tcx| "borrow-checking `{}`", tcx.def_path_str(key.to_def_id()) }
-        cache_on_disk_if(tcx) { tcx.is_typeck_child(key.to_def_id()) }
+        cache_on_disk_if { true }
```

Comment:
Why is this necessary? 

I'm going to do a perf run,? but maybe we just need to check the tainted flag here, too

Reply:
I was getting ICEs in incremental tests complaining that the MIR body for a fn was stolen, since we're now calling borrowck which depends on unoptimized MIR which is probably gone if we do const eval late in the compilation process...

Not sure if only caching if the body if it is tainted works, since we're unconditionally calling borrowck before const eval.

Reply:
Perf does show the additional query result serialization, so we should try to do something here.

Reply:
It's probably not worth it, but we could try adding a query that returns just an `Option<ErrorReported>` and does all the processing of typeck, borrowck and const qualifs. Then we can only cache that on disk and it should rarely get recomputed. But: new query so probably overkill, but worth an experiment?

Reply:
Yeah, I'm not exactly sure what we can do to limit `cache_on_disk` just to when it's needed, since it seems like we need to cache it for any body that has a promoted const, see the failing query stack trace:

```
query stack during panic:
#0 [mir_borrowck] borrow-checking `lit_test::lit_test`
#1 [eval_to_allocation_raw] const-evaluating + checking `lit_test::lit_test::promoted[1]`
#2 [eval_to_allocation_raw] const-evaluating + checking `lit_test::lit_test::promoted[1]`
#3 [optimized_mir] optimizing MIR for `lit_test::lit_test`
#4 [collect_and_partition_mono_items] collect_and_partition_mono_items
```

... but we don't know if a body has any promoted consts until after borrowck.

so I might try the query experiment you suggested. I'll ping you for a perf run when I put it up, since I don't have super powers.

Reply:
> Actually, we do know it at borrowck time, so we could encode that info in the borrowck result.

Oh lol, I thought we did borrowck on `mir_built`, whoopsy.

> When we force mir borrowck and then steal the mir, we own the mir, so we can mutate it and change its tainted field.

So you're suggesting we do add the tainted field to `mir::Body` then, instead of `BorrowCheckerResult`?

Reply:
You need it on both, as (as you reminded me) you can't mutate the mir from borrowck. But at the query that steals the mir after forcing the borrowck query, you can just call borrowck instead of forcing it and copy the tainted field over to the mir body you just stole